circuit_knitting/utils/equivalence.py

```diff
@@ -0,0 +1,133 @@
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Equivalence utilities.
+
+.. currentmodule:: circuit_knitting.utils.equivalence
+
+.. autosummary::
+   :toctree: ../stubs/
+
+"""
+from collections import defaultdict
+
+import numpy as np
+from qiskit.circuit import (
+    EquivalenceLibrary,
+    QuantumCircuit,
+    QuantumRegister,
+    Parameter,
+)
+from qiskit.circuit.library.standard_gates import (
+    RZGate,
+    XGate,
+    YGate,
+    ZGate,
+    HGate,
+    SGate,
+    IGate,
+    SdgGate,
+    SXGate,
+    SXdgGate,
+    TGate,
+    TdgGate,
+    RXGate,
+    RYGate,
+    PhaseGate,
+)
+
+_eagle_sel = HeronEquivalenceLibrary = EagleEquivalenceLibrary = EquivalenceLibrary()
+equivalence_libraries = defaultdict(
+    lambda: None, {"heron": EagleEquivalenceLibrary, "eagle": EagleEquivalenceLibrary}
+)
+
+######################################################################
+
+# YGate
+q = QuantumRegister(1, "q")
+def_y = QuantumCircuit(q)
+for inst in [RZGate(np.pi), XGate()]:
+    def_y.append(inst, [0], [])
+_eagle_sel.add_equivalence(YGate(), def_y)
+
+# ZGate
+q = QuantumRegister(1, "q")
+def_z = QuantumCircuit(q)
+def_z.append(RZGate(np.pi), [0], [])
+_eagle_sel.add_equivalence(ZGate(), def_z)
+
+# HGate
+q = QuantumRegister(1, "q")
+def_h = QuantumCircuit(q)
+for inst in [RZGate(np.pi / 2), SXGate(), RZGate(np.pi / 2)]:
+    def_h.append(inst, [0], [])
+_eagle_sel.add_equivalence(HGate(), def_h)
+
+# SGate
+q = QuantumRegister(1, "q")
+def_s = QuantumCircuit(q)
+def_s.append(RZGate(np.pi / 2), [0], [])
+_eagle_sel.add_equivalence(SGate(), def_s)
+
+# SdgGate
+q = QuantumRegister(1, "q")
+def_sdg = QuantumCircuit(q)
+def_sdg.append(RZGate(-np.pi / 2), [0], [])
+_eagle_sel.add_equivalence(SdgGate(), def_sdg)
+
+# SXdgGate
+q = QuantumRegister(1, "q")
+def_sxdg = QuantumCircuit(q)
+for inst in [
+    RZGate(np.pi / 2),
+    RZGate(np.pi / 2),
+    SXGate(),
+    RZGate(np.pi / 2),
+    RZGate(np.pi / 2),
+]:
+    def_sxdg.append(inst, [0], [])
+_eagle_sel.add_equivalence(SXdgGate(), def_sxdg)
+
+# TGate
+q = QuantumRegister(1, "q")
+def_t = QuantumCircuit(q)
+def_t.append(RZGate(np.pi / 4), [0], [])
+_eagle_sel.add_equivalence(TGate(), def_t)
+
+# TdgGate
+q = QuantumRegister(1, "q")
+def_tdg = QuantumCircuit(q)
+def_tdg.append(RZGate(-np.pi / 4), [0], [])
+_eagle_sel.add_equivalence(TdgGate(), def_tdg)
+
+# RXGate
+q = QuantumRegister(1, "q")
+def_rx = QuantumCircuit(q)
+theta = Parameter("theta")
+for inst in [RZGate(np.pi / 2), SXGate(), RZGate(theta + np.pi), RZGate(5 * np.pi / 2)]:
+    def_rx.append(inst, [0], [])
+_eagle_sel.add_equivalence(RXGate(theta), def_rx)
+
+# RYGate
+q = QuantumRegister(1, "q")
+def_ry = QuantumCircuit(q)
+theta = Parameter("theta")
+for inst in [SXGate(), RZGate(theta + np.pi), SXGate(), RZGate(3 * np.pi)]:
```

Comment:
Should be able to rotate by `pi` at the end here

Reply:
I suspect that the reason there are factors of greater than 2pi in the equivalence library is because a rotation of 2pi doesn't actually bring the wavefunction to its original state; instead, e.g., for [RZGate](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.RZGate), it leads to the wavefunction picking up a global phase of -1.  The rotation must be by an angle of 4pi to bring its state completely to where it began (see also: the mathematics of [spinors](https://en.wikipedia.org/wiki/Spinor)).  Even though a rotation by 2pi leads to a global phase of -1, this will not result in any difference in the state that is actually physically observable, but nonetheless Qiskit carefully keeps track of global phases, and I believe this is one instance of where that leads to some rotation angles that seem a bit atypical.

Here's a quick sanity check (in julia) given the RZGate definition of a rotation by 2pi:

```julia
In [1]: RZ(λ) = [exp(-im * λ / 2) 0; 0 exp(im * λ / 2)]
Out[1]: RZ (generic function with 1 method)

In [2]: RZ(2π)
Out[2]: 2×2 Matrix{ComplexF64}:
 -1.0-1.22465e-16im   0.0+0.0im
  0.0+0.0im          -1.0+1.22465e-16im
```